ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -348,6 +374,186 @@ protected Map<DataNodeId, List<RemoteReplicaInfo>> selectReplicas(Map<DataNodeId
+        groupId++;
+      }
+
+      remoteHostToStandbyNoProgressReplicaGroupId.put(remoteHost, groupId);
+      groupId++;
+    }
+  }
+
+  /**
+   * Generates groups and standby groups and adds these to inflightRemoteReplicaGroup. Removes groups that are done from the
+   * inflightRemoteReplicaGroup. It will generate new groups until shouldTerminateCurrentCycle is false running is true and
+   * maxIterationsPerGroupPerCycle limit is not reached by any iteration count of any group. Replicas will only be added
+   * in the group in groupIdToRemoteReplicaMap and for stand by group in dataNodeIdStandbyReplicasNoProgressReplicaGroupId.
+   * @param groupIdToRemoteReplicaMap {@link Map} mapping of group id to remote replicas
+   * @param dataNodeIdStandbyReplicasNoProgressReplicaGroupId {@link Map} mapping of data node id to standby group id
+   * @param inflightRemoteReplicaGroup {@link Map} map of inflight group id to remote replica groups
+   * @param dataNodeIdToPendingReplicasMarkedForStandByNoProgress {@link Map} map of data node id to standby replicas to pick up
+   * @param remoteReplicaToThrottledTill {@link Map} map of remote replica to time until which it is throttled
+   * @param groupIdIterationCountMap {@link Map} map of group id to number of iterations
+   * @param allReplicasCaughtUpEarly {@link MutableBoolean} stores whether all replicas have caught up early
+   * @param maxIterationsReached {@link MutableBoolean} stores whether max iterations reached
+   */
+  void generateRemoteReplicaGroups(
+      Map<Integer, List<RemoteReplicaInfo>> groupIdToRemoteReplicaMap,
+      Map<DataNodeId, Integer> dataNodeIdStandbyReplicasNoProgressReplicaGroupId,
+      Map<Integer, RemoteReplicaGroup> inflightRemoteReplicaGroup,
+      Map<DataNodeId, Set<RemoteReplicaInfo>> dataNodeIdToPendingReplicasMarkedForStandByNoProgress,
+      Map<RemoteReplicaInfo, Long> remoteReplicaToThrottledTill,
+      Map<Integer, Integer> groupIdIterationCountMap,
+      MutableBoolean allReplicasCaughtUpEarly,
+      MutableBoolean maxIterationsReached
+  ){
+
+    Iterator<Map.Entry<Integer, RemoteReplicaGroup>> inflightRemoteReplicaGroupIterator =
+        inflightRemoteReplicaGroup.entrySet().iterator();
+
+    while (inflightRemoteReplicaGroupIterator.hasNext()) {
+      Map.Entry<Integer, RemoteReplicaGroup> entry = inflightRemoteReplicaGroupIterator.next();
+      RemoteReplicaGroup remoteReplicaGroup = entry.getValue();
+
+      if (remoteReplicaGroup.isDone()) {
+        long currentTime = time.milliseconds();
+        long throttledTill = currentTime + threadThrottleDurationMs;
+        remoteReplicaGroup.remoteReplicaInfos.forEach(remoteReplicaInfo -> {
+          remoteReplicaToThrottledTill.put(remoteReplicaInfo, throttledTill);
+        });
+        storeKeyConverter.tryDropCache(remoteReplicaGroup.getStoreKeysConversionLog());
+        inflightRemoteReplicaGroupIterator.remove();
+      }
+    }
+
```

Comment:
too much whitespace here, and can u add some comments with /** .. */ explaining the code. It's hard to follow if one has no idea of replication